Add invite correlation system

roblabla opened the pull request:
- Allows collaborators (people with the permission to pin stuff) to create single-use invites with the .invite command
- Figure out which invite a user used to join the server.

aveao commented on cogs/invites.py:
```diff
@@ -0,0 +1,33 @@
+    async def invite(self, ctx):
+        welcome_channel = self.bot.get_channel(config.welcome_channel)
+        author = ctx.message.author
+        reason = f"Created by {author.name}#{author.discriminator} ({author.id})"
```
`{author.name}#{author.discriminator}` -> `{str(author)}`

aveao commented on cogs/invites.py:
```diff
@@ -0,0 +1,33 @@
+        try:
+            await ctx.author.send(f"Created single-use invite {invite.url}")
+        except discord.errors.Forbidden:
+            await ctx.send(ctx.author.mention + " I could not send you the invite. Send me a DM so I can reply to you.")
```
Can you put the mention in the string itself with fstrings? It's not that important but it's just how I done it with rest of the code.

aveao commented on cogs/logs.py:
```diff
@@ -28,6 +28,48 @@ def __init__(self, bot):
+        # Add unknown active invites. Can happen if invite was manually created
+        for invite in real_invites:
+            if invite.id not in invites:
+                invites[invite.id] = { "uses": 0, "url": invite.url, "max_uses": invite.max_uses, "code": invite.code }
```
1) This line is over 80 chars, can you split it?
2) Can you remove the spaces after `{` and before `}`?

roblabla commented:
Addressed the style stuff.

I added a gist-based pinboard, unpin command, and handling of pins when we have over 50.

I stress-tested the implementation a bit and... it's kinda slow. Pinning takes a noticeable amount of time (like, ~5 seconds). I didn't profile yet, but I think it's basically taking a long time to download the Gist? It might be a problem, especially if multiple things are pinned in the same channel simultaneously, as it might end up "removing" pins from the pinboard. I wish the Gist API had a way to send patchfiles >_>'.

Maybe we should use raw git, append the links to the local files, and shelling out to git to create and push commits to gist?

aveao commented:
I'm fine with it taking a noticeable time if it gets around to doing it in the end.
